lib/pure/collections/sequtils.nim

```diff
@@ -325,6 +378,61 @@ proc map*[T, S](s: openArray[T], op: proc (x: T): S {.closure.}):
   for i in 0 ..< s.len:
     result[i] = op(s[i])
 
+proc mapIndexed*[T, S](s: openArray[T],
+                       op: proc (i: int, x: T): S {.closure.}): seq[S]{.inline.} =
+  ## Returns a new sequence with the results of `op` proc applied to every
+  ## item together with its index in the container `s`.
+  ##
+  ## Since the input is not modified you can use it to
+  ## transform the type of the elements in the input container.
+  ##
+  ## See also:
+  ## * `mapIt template<#mapIt.t,typed,untyped>`_
+  ## * `apply proc<#apply,openArray[T],proc(T)_2>`_ for the in-place version
+  ##
+  runnableExamples:
+    let
+      a = @[1, 2, 3, 4]
+      b = mapIndexed(a, proc(i: int, x: int): int = x - i)
+    assert b == @[1, 1, 1, 1]
+
+  newSeq(result, s.len)
+  for i, v in s:
+    result[i] = op(i, s[i])
+
+proc scanIndexed*[T](s: openArray[T],
+                     op: proc (i: int, x: T) {.closure.}) {.since: (1, 1).} =
+  ## Apply `op` proc to every item in the container `s`.
+  ##
+  ## See also:
+  ## * `scanIt template<#scanIt.t,typed,untyped>`_
+  ##
+  runnableExamples:
+    let
+      a = @[1, 2, 3, 4]
+    scan(a, proc(x: int) = echo x)
+
+  for i, v in s:
+    op(i, v)
+
+proc scanIndexed*[T, S](s: openArray[T],
+                        op: proc (i: int, x: T): S {.closure.}) {.since: (1, 1).} =
+  ## Apply `op` proc to every item in the container `s`. Results
+  ## of `op` proc are discarded.
```

Comment:
Remove this variant of scanIndexed, it's not how things are done in Nim. Nothing is discarded.

Reply:
I don't with "nothing is to be discarded".

Results are not only for error checking. A case study, procs for drawing entities on a canvas could be:

```nim
proc drawLine(c: Canvas, x, y: float): Canvas = ... return c
proc drawCircle(c: Canvas, x, y, r: float): Canvas = ... return c

# we can put these in a chain:
var c: Canvas = ...
c.drawLine(1, 2).drawCircle(0, 0, 10)
```
Obviously, results of `drawLine`/`drawCircle` can be discarded as we like.

Reply:
Don't lecture me, I said "..are done in Nim". I know the C hack for faking function call chaining really well, thank you.

Reply:
I am not giving lectures, and should not have mentioned those two C functions. Sigh.

Reply:
The point is: Nim APIs are not designed for chaining because for chaining use a 'chain' macro. Do not burden API designers with questions about "chaining" when Nim is pretty strict on ``discard``.